src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -132,6 +158,225 @@ internal void FinalizeDebugInfo()
+
+                // get the type information for the variable
+                ResolvedType? resType = value.QSharpType;
+                DIType? dIType = resType == null ? null : this.DebugTypeFromQsharpType(resType, diBuilder)?.DIType;
+
+                // get the location information for the variable declaration
+                QsLocation? namespaceLoc = this.sharedState.DIManager.CurrentNamespaceElementLocation;
+                Position namespaceOffset = namespaceLoc?.Offset ?? Position.Zero;
+                Position absolutePosition = namespaceOffset + this.TotalOffsetFromStatements() + this.TotalOffsetFromExpressions();
+
+                if (subProgram != null && dIType != null)
+                {
+                    // LLVM Native API has a bug and will crash if we pass in a null dIType
+                    // for either CreateLocalVariable or InsertDeclare (https://bugs.llvm.org/show_bug.cgi?id=52459)
+
+                    // create the debug info for the local variable
+                    DILocalVariable dIVar = diBuilder.CreateLocalVariable(
+                        subProgram,
+                        name,
+                        diBuilder.CompileUnit?.File,
+                        ConvertToDebugPosition(absolutePosition),
+                        dIType,
+                        alwaysPreserve: true,
+                        DebugInfoFlags.None);
+
+                    DILocation dILoc = new DILocation(
+                        this.Context,
+                        ConvertToDebugPosition(absolutePosition),
+                        subProgram);
+
+                    if (this.sharedState.CurrentBlock != null)
+                    {
+                        // create the debug info for the local variable declaration
+                        diBuilder.InsertDeclare(
+                            storage: value.Value,
+                            varInfo: dIVar,
+                            location: dILoc,
+                            insertAtEnd: this.sharedState.CurrentBlock);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates a global function. If the debug flag is set to true, includes debug information
+        /// with the global function.
+        /// </summary>
+        /// <param name="spec">The <see cref="QsSpecialization"/> for the function.</param>
+        /// <param name="mangledName">The mangled name for the function.</param>
+        /// <param name="signature">The signature for the function.</param>
+        /// <param name="isDefinition">Whether this is a function definition (rather than a declaration).</param>
+        /// <returns>The <see cref="IrFunction"/> corresponding to the given Q# type.</returns>
+        internal IrFunction CreateGlobalFunction(QsSpecialization spec, string mangledName, IFunctionType signature, bool isDefinition)
+        {
+            if (this.DebugFlag && spec.Kind == QsSpecializationKind.QsBody)
```

Comment:
Since many of these functions do nothing when `this.DebugFlag` is `false`, to separate this mechanism from other conditions, I would suggest that `this.DebugFlag` is evaluated individually at the beginning of these methods and that the method returns if it is set to false:
```
if (!this.DebugFlag)
{
    return;
}

// Rest of the logic.
```

This is just a suggestion so it is not blocking.